ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,260 @@
-                core_processor::handle_journal(journal, &mut handler);
+            for (program_id, new_state_hash, program_journals) in chunk_journal {
+                // State was updated during journal handling inside the runtime (allocations, pages)
+                in_block_transitions.modify(program_id, |state, _| {
+                    state.hash = new_state_hash;
+                });
+
+                if !program_journals.is_empty() {
+                    for (journal, dispatch_origin) in program_journals {
+                        let mut journal_handler = JournalHandler {
+                            program_id,
+                            dispatch_origin,
+                            controller: TransitionController {
+                                transitions: in_block_transitions,
+                                storage: &db,
+                            },
+                        };
+                        core_processor::handle_journal(journal, &mut journal_handler);
+                    }
+                }
             }
         }
 
-        if no_more_to_do {
+        if chunks.is_empty() {
             break;
         }
     }
-
-    for handle in handles {
-        handle.abort();
-    }
 }
 
-async fn run_task(db: Database, executor: &mut InstanceWrapper, task: Task) {
-    match task {
-        Task::Run {
-            program_id,
-            state_hash,
-            result_sender,
-        } => {
-            let code_id = db.program_code_id(program_id).expect("Code ID must be set");
+// `split_to_chunks` is not exactly sorting (sorting usually `n*log(n)` this one is `O(n)``),
+// but rather partitioning into subsets (chunks) of programs with approximately similar queue sizes.
+fn split_to_chunks(
+    chunk_size: usize,
+    states: &[(ActorId, H256, usize)],
+) -> Vec<Vec<(ActorId, H256, usize)>> {
+    fn chunk_idx(queue_size: usize, number_of_chunks: usize) -> usize {
+        // Simplest implementation of chunk partitioning '| 1 | 2 | 3 | 4 | ..'
+        queue_size.clamp(1, number_of_chunks) - 1
+    }
 
-            let instrumented_code = db.instrumented_code(ethexe_runtime::VERSION, code_id);
+    let number_of_chunks = states.len().div_ceil(chunk_size);
 
-            let journal = executor
-                .run(db, program_id, code_id, state_hash, instrumented_code)
-                .expect("Some error occurs while running program in instance");
+    let mut chunks = Vec::from_iter(iter::repeat_n(Vec::new(), number_of_chunks));
 
-            result_sender.send(journal).unwrap();
-        }
+    for (actor_id, state_hash, queue_size) in states {
+        let chunk_idx = chunk_idx(*queue_size, number_of_chunks);
+        chunks[chunk_idx].push((*actor_id, *state_hash, *queue_size));
     }
+
+    // Merge uneven chunks in reverse order
+    let chunks = chunks.into_iter().flatten().rev().chunks(chunk_size);
+    // Repartition chunks to ensure all chunks have an equal number of elements
```

Comment:
why should the last existing chunk has the maximum amount of elements?

Reply:
Because if for e.x. we have queue sizes [1,2,3,4,5,6,7,8,9,10,12] and we'are trying to split them in 3 chunks:
1. Split them by index - first chunk - [1,], second chunk - [2,], third chunk - [,4,5,6,7,8,9,10,12].
2. Re-partition chunks - [1,2,3,4], [5,6,7,8], [9,10,12]
3. Reverse order - [9,10,12], [5,6,7,8], [1,2,3,4]

Reply:
Also updated docs.

```
//! Chunks are sorted in reverse order (descending), so the first chunk contains the largest queue size.
//! In hypothetic high-load scenarios, this measure may prevent (arguably) starvation of programs with large queue sizes.
//! Also as the entire queue is processed in a single runtime instance in one go, so prioritizing larger queues improves efficiency.
//!
```